mbf_abstract_nav/include/mbf_abstract_nav/abstract_action.h

```diff
@@ -51,106 +50,109 @@ template <typename Action, typename Execution>
 class AbstractAction
 {
  public:
-  typedef boost::bimaps::bimap<uint8_t, std::string> SlotGoalIdMap;
-
   typedef boost::shared_ptr<AbstractAction> Ptr;
 
   typedef typename actionlib::ActionServer<Action>::GoalHandle GoalHandle;
   typedef boost::function<void (GoalHandle &goal_handle, Execution &execution)> RunMethod;
 
+  typedef struct
+  {
+    typename Execution::Ptr execution;
+    boost::thread* thread_ptr;
+    GoalHandle goal_handle;
+  } ConcurrencySlot;
+
   AbstractAction(
       const std::string& name,
       const RobotInformation &robot_info,
       const RunMethod run_method
   ) : name_(name), robot_info_(robot_info), run_(run_method){}
 
-  void start(
-      GoalHandle goal_handle,
+  virtual void start(
+      GoalHandle &goal_handle,
       typename Execution::Ptr execution_ptr
   )
   {
-    boost::lock_guard<boost::mutex> guard(map_mtx_);
-    typename SlotGoalIdMap::left_const_iterator slot
-        = concurrency_slots_.left.find(goal_handle.getGoal()->concurrency_slot);
-    if(slot != concurrency_slots_.left.end()) // if there is a plugin running on the same slot, cancel it // TODO make thread safe
+    uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
+    boost::lock_guard<boost::mutex> guard(slots_mtx_);
+    typename std::map<uint8_t, ConcurrencySlot>::iterator slot_it = concurrency_slots_.find(slot);
+    if(slot_it != concurrency_slots_.end())
     {
-      typename std::map<const std::string, const typename Execution::Ptr>::const_iterator elem
-          = executions_.find(slot->second);
-      if(elem != executions_.end())
-      {
-        elem->second->cancel();
-      }
-      concurrency_slots_.left.erase(slot->first);
+      // if there is a plugin running on the same slot, cancel it
+      slot_it->second.execution->cancel(); // TODO make thread safe
     }
-    concurrency_slots_.insert(SlotGoalIdMap::value_type(goal_handle.getGoal()->concurrency_slot, goal_handle.getGoalID().id));
-    executions_.insert(std::pair<const std::string, const typename Execution::Ptr>(goal_handle.getGoalID().id, execution_ptr));
-    threads_ptrs_.insert(std::pair<const std::string, boost::thread*>(goal_handle.getGoalID().id,
-                         threads_.create_thread(boost::bind(&AbstractAction::runAndCleanUp, this, goal_handle, execution_ptr))));
+    // fill concurrency slot with the new goal handle, execution, and working thread
+    concurrency_slots_[slot].goal_handle = goal_handle;
+    concurrency_slots_[slot].execution = execution_ptr;
+    concurrency_slots_[slot].thread_ptr = threads_.create_thread(boost::bind(&AbstractAction::runAndCleanUp,
+                                                                                 this, goal_handle, execution_ptr));
```

Comment:
If we here write `concurrency_slots_[slot]` instead of `goal_handle` we could avoid the map access in the run method, since it uses the reference version of goal_handle. Thereby it uses the object which lives in the struct. See https://docs.ros.org/melodic/api/actionlib/html/server__goal__handle__imp_8h_source.html#l00340 
Or am I missing something?

Reply:
Not possible, because the references will keep "pointing" to the original GoalHandle (very well explained [here](https://www.walletfox.com/course/pointerref.php)). Should be possible storing and passing a pointer instead of a reference, but then we need to change the _run_ and _runAndCleanUp_ prototypes everywhere

Reply:
what... maybe is not a bad idea after all... :thinking: